packages/workbox-webpack-plugin/src/lib/get-manifest-entries-with-webpack.js

```diff
@@ -0,0 +1,137 @@
+/*
+  Copyright 2017 Google Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+const resolveWebpackUrl = require('./utils/resolve-webpack-url');
+
+/**
+ * A single manifest entry that workbox-sw can precache
+ *
+ * @param {string} url Webpack asset url path
+ * @param {string} [revision] A revision hash for the entry
+ * @return {module:workbox-build.ManifestEntry} A single manifest entry
+ *
+ * @private
+ */
+const entry = (url, revision) => (revision
+  ? {url, revision}
+  : {url}
+);
+
+/**
+ * Filter chunks that have one of the provided names (from config.chunks)
+ *
+ * TODO:
+ *  Filter files by size:
+ *    https://github.com/GoogleChrome/workbox/pull/808#discussion_r139606242
+ *  Filter files that match `staticFileGlobsIgnorePatterns` (or something)
+ *  but filter for [/\.map$/, /asset-manifest\.json$/] by default:
+ *    https://github.com/GoogleChrome/workbox/pull/808#discussion_r140565156
+ *
+ * @param {Array<Object>} chunks Webpack chunks
+ * @param {Array<string>} include Chunk names to include
+ * @param {Array<string>} exclude Chunk names to exclude
+ * @return {Array<Object>} Filtered array of chunks
+ *
+ * @private
+ */
+const filterChunks = (chunks, include, exclude) =>
+  chunks.filter((chunk) =>
+    // chunk has a name
+    Object.prototype.hasOwnProperty.call(chunk, 'name')
+      // did not specify include or chunk in include
+      && (!Array.isArray(include) || !!~include.indexOf(chunk.name))
+      // did not specify exclude or chunk not in exclude
+      && (!Array.isArray(exclude) || !~exclude.indexOf(chunk.name))
+  );
+
+/**
+ * Maps webpack asset filenames to their chunk hash
+ *
+ * TODO:
+ *   Elaborate on this function description:
+ *      https://github.com/GoogleChrome/workbox/pull/808#discussion_r139605066
+ *
+ * @param {Array<Object>} chunks Webpack chunks
+ * @return {Object} {filename: hash} map
+ *
+ * @private
+ */
+const mapAssetsToChunkHash = (chunks) =>
+  chunks.reduce((assetMap, chunk) =>
+      Object.assign(
+        assetMap,
+        ...chunk.files.map((f) => ({[f]: chunk.renderedHash}))
+    ), {});
+
+/**
+ * Generate an array of manifest entries using webpack's compilation data
```

Comment:
Question: For dev environments, should workbox filter out source maps? Also, some people may be creating source maps even for production builds to debug deployed code. Some users may need to filter out some assets depending on their use cases.

Reply:
See https://github.com/GoogleChrome/workbox/pull/808/files/a2dcaf4235f9f10456988a78759eb17dfedf6823#diff-dff2a0e4847723c3a1d299d74db40c0cR39

The idea was to provide a config option (`assetIgnoreRegex` or something) where users can provide an array of regular expressions to filter asset filenames out. By default we would use: [/\.map$/, /asset-manifest\.json$/]

It feels a little odd filtering by asset filenames. Normally (in my experience) webpack configs include/exclude assets based on chunk names, not actual filenames but for the case with `*.map$` files, they would need to be filtered by asset file names.

Reply:
I guess source-maps can be filtered out by default. However I'm not sure how that would affect development experience when testing offline caching. Need to test it.

This problem is similar to how loaders handle file inclusion/exclusion using three properties: `{test, include, exclude}` and the code for that is in https://github.com/webpack/webpack/blob/master/lib/ModuleFilenameHelpers.js, look for `matchObject`. That's what `uglifyjs-webpack-plugin` is also using in  https://github.com/webpack-contrib/uglifyjs-webpack-plugin/blob/master/src/index.js.

Reply:
So considering we filter both **chunkname** and **filename** what should the api look like?

We could provide explicit chunk / filename filtering eg:
```js
WorkboxPlugin({
  ...
  filterChunks: {
    test: /\.js($&#124;\?)/i,
  },
  filterFilenames: {
    exclude: /\/excludes/,
  },
});
```

Or we can just filter both using the same properties, eg:
```js
WorkboxPlugin({
  ...
  test: /\.js($&#124;\?)/i,
  exclude: /\/excludes/,
});
```
and just filter both **chunknames** and **filenames** by those provided values?

Something else to consider is [html-webpack-plugin's config](https://github.com/jantimon/html-webpack-plugin#configuration) the `{chunks, excludeChunks}` options are very nice. Maybe use those and apply the `{test, include, exclude}` options *only* to **filenames** (**chunknames** are handled by the `chunks` option...

Reply:
@anilanar ^ (meant to tag you in that comment)

Reply:
FYI: I've opened https://github.com/GoogleChrome/workbox/issues/935 to track this, and moved over some relevant conversation. I don't want too much of the conversation to be lost in a closed PR.